vllm/entrypoints/chat_utils.py

```diff
@@ -668,6 +738,69 @@ def parse_chat_messages_futures(
     return conversation, mm_tracker.all_mm_data()
 
 
+def _is_var_access(node: jinja2.nodes.Node, varname: str) -> bool:
+    if isinstance(node, jinja2.nodes.Name):
+        return node.ctx == "load" and node.name == varname
+
+    return False
+
+
+def _is_attr_access(node: jinja2.nodes.Node, varname: str, key: str) -> bool:
+    if isinstance(node, jinja2.nodes.Getitem):
+        return (node.ctx == "load" and _is_var_access(node.node, varname)
+                and isinstance(node.arg, jinja2.nodes.Const)
+                and node.arg.value == key)
+
+    if isinstance(node, jinja2.nodes.Getattr):
+        return (node.ctx == "load" and _is_var_access(node.node, varname)
+                and node.attr == key)
+
+    return False
+
+
+def _iter_nodes_define_message(chat_template_ast: jinja2.nodes.Template):
+    # Search for {%- for message in messages -%} loops
+    for loop_ast in chat_template_ast.find_all(jinja2.nodes.For):
+        loop_iter = loop_ast.iter
+        loop_target = loop_ast.target
+
+        if _is_var_access(loop_iter, "messages"):
+            assert isinstance(loop_target, jinja2.nodes.Name)
+            yield loop_ast, loop_target.name
+
+
+def _iter_nodes_define_content_item(chat_template_ast: jinja2.nodes.Template):
+    for node, message_varname in _iter_nodes_define_message(chat_template_ast):
+        # Search for {%- for content in message['content'] -%} loops
+        for loop_ast in node.find_all(jinja2.nodes.For):
+            loop_iter = loop_ast.iter
+            loop_target = loop_ast.target
+
+            if _is_attr_access(loop_iter, message_varname, "content"):
+                assert isinstance(loop_target, jinja2.nodes.Name)
+                yield loop_iter, loop_target.name
```

Comment:
This handles the common case of iterating through "openai" formatted `message['content']` as a list. Please tell me if you are aware of any chat templates that expects `message['content']` as a list but doesn't iterate through it for some reason.